module/extension/Director/src/Director.cpp

```diff
@@ -0,0 +1,105 @@
+#include "Director.h"
+
+#include "extension/Configuration.h"
+
+namespace module {
+namespace extension {
+
+    using extension::Configuration;
+
+    Director::Director(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+
+        on<Configuration>("Director.yaml").then([this](const Configuration& config) {
+            // Use configuration here from file Director.yaml
+        });
+
+        on<Trigger<EnteringReaction>>().then([this](const EnteringReaction& r) {
+            // TODO add an entering flag for this provider
+        });
+
+        on<Trigger<LeavingReaction>>().then([this](const LeavingReaction& r) {
+            // TODO add a leaving flag for this provider
+        });
+
+        on<Trigger<ProvidesReaction>>().then([this](const ProvidesReaction& r) {
+            // TODO add a provider flag for this provider
+        });
+
+        on<Trigger<WhenExpression>>().then([this](const WhenReaction& r) {
+            // TODO add a when expression for this provider
+        });
+
+        on<Trigger<CausingExpression>>().then([this](const CausingReaction& r) {
+            // TODO add a causing expression for this provider
+        });
+
+        on<Trigger<DirectorTask>>().then([this](const DirectorTask& task) {
+            // Check which provider emitted this task (if any) to get its scope
+
+            // TODO if the cause_id is not any of our reactions remove it from scope
+        });
+
+        // Main director function
+        on<Every<90, Per<std::chrono::seconds>>>().then([this] {
+            // Sort tasks by priority
+            std::sort(global_tasks.begin(), global_tasks.end(), bypriority);
+
+            for (const auto& t : global_tasks) {
+
+                // We should have a current state tree (which initially may be empty), a transitional state tree and a
```

Comment:
This should appear in a separate document in the doc folder with some diagrams to explain the steps and conditions in this algorithm.

Reply:
This is nowhere near fleshed out enough for that yet